_posts/2023-06-13-http4s-authentication-part2.md

```diff
@@ -0,0 +1,1032 @@
+  private val secret: Array[Byte] =
+    Properties.envOrElse("SECRET_KEY", "").getBytes()
+}
+```
+
+We create `OtpService` as a class in order to pass a `Generator` and `User` as constructor arguments during instantiation. The `secret` value is accessed by calling `Properties.envOrElse()`, if the environmental variable is not present the function assigns an empty string, and the `getBytes()` stores the `secret` as an `Array[Byte]`.
+
+```scala
+import com.bastiaanjansen.otp.*
+import java.time.Duration
+
+class OtpService(generator: Generator, user: User){
+    ...
+ private object GenInstance {
+    def hotpGen(secretValue: Array[Byte]) =
+      new HOTPGenerator.Builder(secretValue)
+        .withPasswordLength(6)
+        .withAlgorithm(HMACAlgorithm.SHA256)
+        .build()
+    def totpGen(secretValue: Array[Byte]) =
+      new TOTPGenerator.Builder(secretValue)
+        .withHOTPGenerator(builder => {
+          builder.withPasswordLength(6)
+          builder.withAlgorithm(HMACAlgorithm.SHA256)
+        })
+        .withPeriod(Duration.ofSeconds(30))
+        .build()
+  }
+}
+```
+
+The `GenInstance` object contains two functions `hotpGen()` and `totpGen()` used to provide instances of `HOTPGenerator` and `TOTPGenerator` respectively, they both take an `Array[Byte]` as a parameter. Note that Google Authenticator displays a token length of 6 therefore we supply the same value to our function implementations. We also use the same algorithm type, `SHA256` that we defined in our `URI` string. This object is created as `private` so that it's only used within this class.
+
+```scala
+import cats.effect.IO
+import com.rockthejvm.Generator.*
+
+class OtpService(generator: Generator, user: User){
+    ...
+  def getToken: IO[String] =
+    IO {
+      generator match
+        case Hotp =>
+          GenInstance.hotpGen(secret).generate(user.getCounter)
+        case Totp =>
+          GenInstance.totpGen(secret).now()
+    }
+}
+```
+
+The `getToken()` function generates a token and returns an `IO[String]` depending on the type of generator supplied. It uses arguments from the class constructor in its implementation.
+
+```scala
+class OtpService(generator: Generator, user: User){
+    ...
+  def verifyCode(code: String): IO[Boolean] =
+    IO {
+      generator match
+        case Hotp =>
+          GenInstance
+            .hotpGen(secret)
+            .verify(code, user.getCounter)
+        case Totp =>
+          GenInstance
+            .totpGen(secret)
+            .verify(code)
+    }
+}
+```
+
+The `verifyCode()` function is used to check if the code the user sends to our service is genuine. It takes the `code` (the otp token) and counter value `user.getCounter()` for the case of `Hotp` and returns an `IO[Boolean]` which shows if verification has failed or passed.
+
+```scala
+class OtpService(generator: Generator, user: User){
+  ...
+  def makeBarCodeImg: IO[Unit] = BarCodeService.createQRCode(
```

Comment:
maybe pass name and email as arguments here instead of hardcoding them in the impl